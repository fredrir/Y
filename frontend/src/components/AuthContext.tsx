import React, { createContext, useContext, useState, useEffect } from "react";
import { jwtDecode } from "jwt-decode";

interface AuthContextType {
  isLoggedIn: boolean;
  token: string | null;
  user: { id: string; username: string } | null;
  login: (token: string) => void;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

interface DecodedToken {
  id: string;
  username: string;
  iat: number;
  exp: number;
}

export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [token, setToken] = useState<string | null>(null);
  const [user, setUser] = useState<{ id: string; username: string } | null>(
    null,
  );

  useEffect(() => {
    const storedToken = localStorage.getItem("token");
    if (storedToken) {
      try {
        const decoded: DecodedToken = jwtDecode(storedToken);
        // Check if token is expired
        if (decoded.exp * 1000 < Date.now()) {
          // Token expired
          logout();
        } else {
          setIsLoggedIn(true);
          setToken(storedToken);
          setUser({ id: decoded.id, username: decoded.username });
        }
      } catch (error) {
        console.error("Invalid token:", error);
        logout();
      }
    }
  }, []);

  const login = (newToken: string) => {
    try {
      const decoded: DecodedToken = jwtDecode(newToken);
      localStorage.setItem("token", newToken);
      setIsLoggedIn(true);
      setToken(newToken);
      setUser({ id: decoded.id, username: decoded.username });
    } catch (error) {
      console.error("Invalid token:", error);
    }
  };

  const logout = () => {
    localStorage.removeItem("token");
    setIsLoggedIn(false);
    setToken(null);
    setUser(null);
  };

  return (
    <AuthContext.Provider value={{ isLoggedIn, token, user, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
