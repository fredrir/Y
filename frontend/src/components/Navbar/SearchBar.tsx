import useDebounce from "@/hooks/useDebounce";
import { HashtagType, UserType } from "@/lib/types";
import { SEARCH_HASHTAGS, SEARCH_USERS } from "@/queries/search";
import { useQuery } from "@apollo/client";
import { ChangeEvent, KeyboardEvent, useEffect, useRef, useState } from "react";
import { Popover, PopoverContent, PopoverTrigger } from "../ui/popover";
import Username from "../Username";

const SearchBar = () => {
  const params = new URLSearchParams(location.search);
  const currentQuery = params.get("q") || "";
  const [searchQuery, setSearchQuery] = useState(currentQuery);
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [activeSuggestionIndex, setActiveSuggestionIndex] = useState(0);
  const [suggestions, setSuggestions] = useState<(HashtagType | UserType)[]>(
    [],
  );
  const inputRef = useRef<HTMLInputElement>(null);
  const popoverRef = useRef<HTMLDivElement>(null);

  const debouncedQuery = useDebounce(searchQuery, 300);

  const {
    data: hashtagsData,
    loading: hashtagsLoading,
    refetch: refetchHashtags,
  } = useQuery<{ searchHashtags: HashtagType[] }>(SEARCH_HASHTAGS, {
    variables: { query: debouncedQuery, page: 1, limit: 5 },
  });

  const {
    data: mentionsData,
    loading: mentionsLoading,
    refetch: refetchUsers,
  } = useQuery<{ searchUsers: UserType[] }>(SEARCH_USERS, {
    variables: { query: debouncedQuery, page: 1, limit: 5 },
  });

  useEffect(() => {
    refetchHashtags();
    refetchUsers();
  }, [debouncedQuery, refetchHashtags, refetchUsers]);

  useEffect(() => {
    setSuggestions([
      ...(hashtagsData?.searchHashtags ?? []),
      ...(mentionsData?.searchUsers ?? []),
    ]);
  }, [hashtagsData, mentionsData]);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        inputRef.current &&
        !inputRef.current.contains(event.target as Node) &&
        popoverRef.current &&
        !popoverRef.current.contains(event.target as Node)
      ) {
        setShowSuggestions(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  const handleSearch = () => {
    if (activeSuggestionIndex === 0 || suggestions.length === 0) {
      window.location.href = `/project2/search?q=${encodeURIComponent(
        searchQuery,
      )}`;
      return;
    }
    const selectedSuggestion = suggestions[activeSuggestionIndex - 1];
    window.location.href = `/project2/${
      selectedSuggestion.__typename === "User"
        ? `user/${selectedSuggestion.username}`
        : `hashtag/${selectedSuggestion.tag}`
    }`;
  };

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(e.target.value);
    if (!showSuggestions) {
      setShowSuggestions(true);
      setActiveSuggestionIndex(0);
    }
  };

  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      e.preventDefault();
      handleSearch();
      return;
    }
    const maxIndex = suggestions.length > 5 ? 5 : suggestions.length;
    if (showSuggestions) {
      if (e.key === "ArrowDown") {
        e.preventDefault();
        setActiveSuggestionIndex((prevIndex) =>
          prevIndex === maxIndex ? 0 : prevIndex + 1,
        );
      } else if (e.key === "ArrowUp") {
        e.preventDefault();
        setActiveSuggestionIndex((prevIndex) =>
          prevIndex === 0 ? maxIndex : prevIndex - 1,
        );
      }
    }
  };
  return (
    <Popover open={showSuggestions}>
      <PopoverTrigger asChild>
        <input
          ref={inputRef}
          type="search"
          id="search"
          maxLength={40}
          placeholder="Search here..."
          autoComplete="off"
          className="w-full max-w-xs rounded-md bg-gray-100 p-2 outline-none dark:bg-gray-800"
          value={searchQuery}
          onKeyDown={handleKeyDown}
          onChange={handleInputChange}
          onClick={() => setShowSuggestions(true)}
          onMouseEnter={() => setActiveSuggestionIndex(0)}
        />
      </PopoverTrigger>
      {suggestions.length > 0 && (
        <PopoverContent
          ref={popoverRef}
          onOpenAutoFocus={(e) => e.preventDefault()}
          className="z-[70] overflow-hidden p-0"
          align="start"
        >
          {hashtagsLoading || mentionsLoading ? (
            <p className="w-full p-2">Loading...</p>
          ) : (
            <div className="flex w-full appearance-none flex-col overflow-hidden outline-none">
              {suggestions.slice(0, 5).map((suggestion, index) => {
                const isUser = suggestion.__typename === "User";
                return (
                  <a
                    key={isUser ? suggestion.id : suggestion.tag}
                    className={`flex w-full cursor-pointer items-center gap-1 p-2 ${
                      index + 1 === activeSuggestionIndex
                        ? "bg-blue-500 text-white dark:bg-blue-800"
                        : ""
                    }`}
                    href={`/project2/${
                      isUser
                        ? `user/${suggestion.username}`
                        : `hashtag/${suggestion.tag}`
                    }`}
                    onMouseEnter={() => setActiveSuggestionIndex(index + 1)}
                  >
                    {isUser ? (
                      <Username
                        noHref
                        user={suggestion}
                        customBadgeColors={
                          index + 1 === activeSuggestionIndex
                            ? "text-white"
                            : "text-blue-500 dark:text-blue-400"
                        }
                      />
                    ) : (
                      <div className="flex items-center gap-1">
                        <span className="flex size-8 h-full items-center justify-center text-2xl">
                          #
                        </span>
                        <p>{suggestion.tag}</p>
                      </div>
                    )}
                  </a>
                );
              })}
            </div>
          )}
        </PopoverContent>
      )}
    </Popover>
  );
};

export default SearchBar;
