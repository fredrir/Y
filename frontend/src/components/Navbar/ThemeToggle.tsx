import { useEffect, useState } from "react";
import { Sun, Moon } from "lucide-react";

const ThemeToggle = () => {
  const [theme, setTheme] = useState(() => {
    if (typeof window !== "undefined" && window.localStorage) {
      const storedTheme = window.localStorage.getItem("theme");
      return storedTheme ? storedTheme : "light";
    }
    return "light";
  });

  useEffect(() => {
    const root = document.documentElement;
    root.classList.toggle("dark", theme === "dark");

    if (typeof window !== "undefined" && window.localStorage) {
      window.localStorage.setItem("theme", theme);
    }
  }, [theme]);

  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === "dark" ? "light" : "dark"));
  };

  return (
    <>
      <div className="hidden md:block">
        <button
          onClick={toggleTheme}
          className={`mx-4 h-10 w-16 rounded-full p-1 transition-colors duration-300 focus:outline-none ${
            theme === "dark" ? "bg-gray-900" : "bg-gray-300"
          }`}
          aria-label={
            theme === "dark" ? "Switch to light theme" : "Switch to dark theme"
          }
        >
          <div
            className={`flex h-8 w-8 transform items-center justify-center rounded-full transition-transform duration-300 ${
              theme === "dark"
                ? "translate-x-6 bg-gray-800"
                : "translate-x-0 bg-white"
            }`}
          >
            {theme === "dark" ? (
              <Moon className="h-5 w-5 text-yellow-300" />
            ) : (
              <Sun className="h-5 w-5 text-yellow-500" />
            )}
          </div>
        </button>
      </div>
      <div className="md:hidden">
        <button
          className="flex w-full flex-row items-center gap-2 p-2 text-left text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
          onClick={toggleTheme}
        >
          {theme === "dark" ? (
            <Sun className="h-5 w-5" />
          ) : (
            <Moon className="h-5 w-5" />
          )}
          {theme === "dark" ? <p>Light Mode</p> : <p>Dark Mode</p>}
        </button>
      </div>
    </>
  );
};

export default ThemeToggle;
