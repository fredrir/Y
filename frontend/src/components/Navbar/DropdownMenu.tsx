import { useAuth } from "@/components/AuthContext";
import ThemeToggle from "@/components/Navbar/ThemeToggle";
import { UsersIcon } from "@heroicons/react/24/outline";
import {
  Bars3Icon,
  HashtagIcon,
  HomeIcon,
  UserIcon,
  XMarkIcon,
} from "@heroicons/react/24/outline";
import { LogInIcon, LogOutIcon } from "lucide-react";
import { useEffect, useRef, useState } from "react";

export const DropdownMenu = () => {
  const [isOpen, setIsOpen] = useState(false);
  const menuRef = useRef<HTMLDivElement>(null);
  const { user, logout } = useAuth();

  const routes = [
    {
      name: user ? "Profile" : "Login",
      href: `/project2/${user ? `user/${user.username}` : "login"}`,
      icon: user ? (
        <UserIcon className="size-5" />
      ) : (
        <LogInIcon className="size-5" />
      ),
    },
    {
      name: "Homepage",
      href: "/project2",
      icon: <HomeIcon className="size-5" />,
    },
    {
      name: "Users",
      href: "/project2/users",
      icon: <UsersIcon className="size-5" />,
    },
    {
      name: "Trending",
      href: "/project2/hashtag",
      icon: <HashtagIcon className="size-5" />,
    },
  ];

  const toggleMenu = () => {
    setIsOpen(!isOpen);
  };

  const closeMenu = () => {
    setIsOpen(false);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [menuRef]);

  return (
    <div className="relative block lg:hidden" ref={menuRef}>
      <div
        onClick={toggleMenu}
        className="cursor-pointer transition-transform duration-300"
      >
        <div
          className={`transition-transform duration-300 ${
            isOpen ? "rotate-180" : ""
          }`}
        >
          {isOpen ? (
            <XMarkIcon className="size-8" />
          ) : (
            <Bars3Icon className="size-8" />
          )}
        </div>
      </div>

      <div
        className={`absolute right-0 top-12 z-50 mt-2 w-48 overflow-hidden rounded-md border border-gray-500 bg-gray-300 text-gray-900 shadow-lg transition-all duration-300 dark:bg-gray-800 dark:text-white ${
          isOpen
            ? "scale-100 opacity-100"
            : "pointer-events-none scale-95 opacity-0"
        }`}
      >
        {routes.map((route, i: number) => (
          <a
            href={route.href}
            key={i}
            onClick={closeMenu}
            className="flex flex-row items-center gap-2 p-2 text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
          >
            {route.icon}
            {route.name}
          </a>
        ))}
        <ThemeToggle />
        {user && (
          <button
            onClick={() => {
              logout();
              closeMenu();
            }}
            className="flex w-full flex-row items-center gap-2 p-2 text-left text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
          >
            <LogOutIcon className="size-5" />
            Log out
          </button>
        )}
      </div>
    </div>
  );
};
