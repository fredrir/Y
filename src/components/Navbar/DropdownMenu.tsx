import { useState, useRef, useEffect } from "react";
import { Bars3Icon, XMarkIcon } from "@heroicons/react/24/outline";
import ThemeToggle from "./ThemeToggle";

export const DropdownMenu = () => {
  const [isOpen, setIsOpen] = useState(false);
  const menuRef = useRef<HTMLDivElement>(null);

  const routes = [
    { name: "Profile", href: "/" },
    { name: "Homepage", href: "/" },
  ];

  const toggleMenu = () => {
    setIsOpen(!isOpen);
  };

  const closeMenu = () => {
    setIsOpen(false);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (menuRef.current && !menuRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener("mousedown", handleClickOutside);

    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [menuRef]);

  return (
    <div className="relative block md:hidden" ref={menuRef}>
      <div
        onClick={toggleMenu}
        className="cursor-pointer transition-transform duration-300"
      >
        <div
          className={`transition-transform duration-300 ${
            isOpen ? "rotate-180" : ""
          }`}
        >
          {isOpen ? (
            <Bars3Icon className="text-white h-8 w-8" />
          ) : (
            <XMarkIcon className="text-white h-8 w-8" />
          )}
        </div>
      </div>

      <div
        className={`absolute right-0 mt-2 top-12 w-48 z-50 bg-gray-300 dark:bg-gray-800 border border-gray-500 text-gray-900 dark:text-white  rounded-md shadow-lg transition-all duration-300 ${
          isOpen
            ? "opacity-100 scale-100"
            : "opacity-0 scale-95 pointer-events-none"
        }`}
      >
        {routes.map((route) => (
          <a
            href={route.href}
            key={route.href}
            onClick={closeMenu}
            className="block px-4 py-2 text-lg hover:bg-gray-500 dark:hover:bg-gray-900"
          >
            {route.name}
          </a>
        ))}
        <ThemeToggle />
      </div>
    </div>
  );
};
