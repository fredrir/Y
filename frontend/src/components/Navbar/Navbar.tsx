import { useAuth } from "@/components/AuthContext";
import { DropdownMenu } from "@/components/Navbar/DropdownMenu";
import ThemeToggle from "@/components/Navbar/ThemeToggle";
import Avatar from "@/components/Profile/Avatar";
import useDebounce from "@/hooks/useDebounce";
import { HashtagType, UserType } from "@/lib/types";
import { SEARCH_HASHTAGS, SEARCH_USERS } from "@/queries/search";
import { useQuery } from "@apollo/client";
import { ChangeEvent, KeyboardEvent, useEffect, useState } from "react";
import { Popover, PopoverContent, PopoverTrigger } from "../ui/popover";

const Navbar = () => {
  const params = new URLSearchParams(location.search);
  const currentQuery = params.get("q") || "";
  const [searchQuery, setSearchQuery] = useState(currentQuery);
  const { logout, user } = useAuth();
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [activeSuggestionIndex, setActiveSuggestionIndex] = useState(0);
  const [suggestions, setSuggestions] = useState<(HashtagType | UserType)[]>(
    [],
  );

  const debouncedQuery = useDebounce(searchQuery, 300);

  const {
    data: hashtagsData,
    loading: hashtagsLoading,
    refetch: refetchHashtags,
  } = useQuery<{
    searchHashtags: HashtagType[];
  }>(SEARCH_HASHTAGS, {
    variables: { query: debouncedQuery, page: 1, limit: 5 },
    skip: debouncedQuery.length < 1,
  });

  const {
    data: mentionsData,
    loading: mentionsLoading,
    refetch: refetchUsers,
  } = useQuery<{
    searchUsers: UserType[];
  }>(SEARCH_USERS, {
    variables: { query: debouncedQuery, page: 1, limit: 5 },
    skip: debouncedQuery.length < 1,
  });

  useEffect(() => {
    if (debouncedQuery.length > 0) {
      refetchHashtags();
      refetchUsers();
    }
  }, [debouncedQuery, refetchHashtags, refetchUsers]);

  useEffect(() => {
    setSuggestions([
      ...(hashtagsData?.searchHashtags ?? []),
      ...(mentionsData?.searchUsers ?? []),
    ]);
  }, [hashtagsData, mentionsData]);

  const handleSearch = () => {
    if (activeSuggestionIndex === 0 || suggestions.length === 0) {
      window.location.href = `/project2/search?q=${encodeURIComponent(
        searchQuery,
      )}`;
      return;
    }
    const selectedSuggestion = suggestions[activeSuggestionIndex - 1];
    window.location.href = `/project2/${
      selectedSuggestion.__typename === "User"
        ? `user/${selectedSuggestion.username}`
        : `hashtag/${selectedSuggestion.tag}`
    }`;
  };

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(e.target.value);
    if (!showSuggestions) {
      setShowSuggestions(true);
      setActiveSuggestionIndex(0);
    }
  };

  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === "Enter") {
      e.preventDefault();
      handleSearch();
      return;
    }
    if (showSuggestions) {
      if (e.key === "ArrowDown") {
        e.preventDefault();
        setActiveSuggestionIndex((prevIndex) =>
          prevIndex === suggestions.length - 2 ? 0 : prevIndex + 1,
        );
      } else if (e.key === "ArrowUp") {
        e.preventDefault();
        setActiveSuggestionIndex((prevIndex) =>
          prevIndex === 0 ? suggestions.length - 2 : prevIndex - 1,
        );
      }
    }
  };

  const login = () => {
    window.location.href = "/project2/login";
  };

  return (
    <>
      <nav className="fixed z-[60] flex h-20 w-full items-center justify-between bg-gray-200/40 px-5 py-5 backdrop-blur-sm dark:bg-gray-950/80">
        <a
          href="/project2"
          className="group flex items-center justify-center gap-5 hover:scale-110 hover:text-gray-300"
        >
          <h1 className="sm:text-2x text-2xl font-semibold lg:text-2xl xl:text-3xl">
            <span className="group-hover:text-gray-600 group-hover:dark:text-gray-300">
              Y
            </span>
          </h1>
        </a>

        <div className="mx-4 flex max-w-xs flex-1 items-center justify-center gap-2 sm:max-w-lg lg:justify-end">
          <Popover open={showSuggestions}>
            <PopoverTrigger asChild>
              <input
                type="search"
                id="search"
                maxLength={40}
                placeholder="Search here..."
                autoComplete="off"
                className="w-full rounded-md bg-gray-100 p-2 outline-none dark:bg-gray-800"
                value={searchQuery}
                onKeyDown={handleKeyDown}
                onChange={handleInputChange}
                onMouseEnter={() => setActiveSuggestionIndex(0)}
              />
            </PopoverTrigger>
            {suggestions.length > 0 && (
              <PopoverContent
                onOpenAutoFocus={(e) => e.preventDefault()}
                className="z-[70] overflow-hidden p-0"
              >
                {hashtagsLoading || mentionsLoading ? (
                  <p className="w-full p-2">Loading...</p>
                ) : (
                  <div className="flex w-full appearance-none flex-col overflow-hidden outline-none">
                    {suggestions.slice(0, 5).map((suggestion, index) => {
                      const isUser = suggestion.__typename === "User";
                      return (
                        <a
                          key={isUser ? suggestion.id : suggestion.tag}
                          className={`flex w-full cursor-pointer items-center gap-2 p-2 ${
                            index + 1 === activeSuggestionIndex
                              ? "bg-blue-500 text-white dark:bg-blue-800"
                              : ""
                          }`}
                          href={`/project2/${
                            isUser
                              ? `user/${suggestion.username}`
                              : `hashtag/${suggestion.tag}`
                          }`}
                          onMouseEnter={() =>
                            setActiveSuggestionIndex(index + 1)
                          }
                        >
                          {isUser ? (
                            <Avatar noHref user={suggestion} />
                          ) : (
                            <p className="flex size-8 h-full items-center justify-center text-2xl">
                              #
                            </p>
                          )}

                          <p>
                            {isUser
                              ? `${suggestion.username}`
                              : `${suggestion.tag}`}
                          </p>
                        </a>
                      );
                    })}
                  </div>
                )}
              </PopoverContent>
            )}
          </Popover>
          <div className="hidden min-w-fit items-center gap-2 lg:flex">
            <ThemeToggle />
            <div className="flex items-center gap-2">
              {user && (
                <div className="mr-2">
                  <Avatar user={user} large={false} />{" "}
                </div>
              )}
              {user ? (
                <button
                  onClick={logout}
                  className="rounded-md bg-red-500 px-3 py-2 text-sm font-semibold text-white hover:bg-red-600 dark:bg-red-600 dark:hover:bg-red-700"
                >
                  Log out
                </button>
              ) : (
                <button
                  onClick={login}
                  className="rounded-md bg-green-500 px-3 py-2 text-sm font-semibold text-white hover:bg-green-600 dark:bg-green-600 dark:hover:bg-green-700"
                >
                  Log in
                </button>
              )}
            </div>
          </div>
        </div>
        <DropdownMenu />
      </nav>
      <div className="h-20"></div>
    </>
  );
};

export default Navbar;
