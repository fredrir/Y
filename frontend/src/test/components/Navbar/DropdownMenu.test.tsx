import { DropdownMenu } from "@/components/Navbar/DropdownMenu";
import { useAuth } from "@/hooks/AuthContext";
import { MockedProvider } from "@apollo/client/testing";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";

vi.mock("@/hooks/AuthContext", () => ({
  useAuth: vi.fn().mockReturnValue({
    user: { username: "testuser" },
    logout: vi.fn(),
  }),
}));

describe("DropdownMenu", () => {
  it("matches the snapshot", () => {
    const { container } = render(
      <MockedProvider>
        <DropdownMenu />
      </MockedProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the dropdown toggle Bars3Icon by default", () => {
    render(<DropdownMenu />);

    const barsIcon = screen.getByTestId("bars-icon");
    expect(barsIcon).toBeInTheDocument();
    expect(barsIcon).toBeVisible();
  });

  it("toggles the dropdown when clicked, showing XMarkIcon when open", async () => {
    render(<DropdownMenu />);

    let barsIcon = screen.getByTestId("bars-icon");
    expect(barsIcon).toBeInTheDocument();
    expect(barsIcon).toBeVisible();

    await userEvent.click(barsIcon);

    const xMarkIcon = screen.getByTestId("xmark-icon");
    expect(xMarkIcon).toBeInTheDocument();
    expect(xMarkIcon).toBeVisible();

    await userEvent.click(xMarkIcon);

    barsIcon = screen.getByTestId("bars-icon");
    expect(barsIcon).toBeInTheDocument();
    expect(barsIcon).toBeVisible();
  });

  it("renders the correct routes", () => {
    render(<DropdownMenu />);

    const routes = [
      { name: "Profile", href: "/user/testuser" },
      { name: "Homepage", href: "/" },
      { name: "Users", href: "/users" },
      { name: "Trending", href: "/hashtag" },
    ];

    routes.forEach((route) => {
      const routeLink = screen.getByText(route.name);
      expect(routeLink).toBeInTheDocument();
      expect(routeLink.closest("a")).toHaveAttribute("href", route.href);
    });
  });

  it("renders the logout button for logged-in users", () => {
    render(<DropdownMenu />);
    const logoutButton = screen.getByText(/Log out/i);
    expect(logoutButton).toBeInTheDocument();
  });

  it("renders the correct routes when user is logged out", () => {
    vi.mocked(useAuth).mockReturnValue({
      logout: vi.fn(),
      isLoggedIn: false,
      token: null,
      user: null,
      login: vi.fn(),
      following: [],
      setFollowing: vi.fn(),
      refetchUser: vi.fn(),
    });

    render(<DropdownMenu />);

    // Adjust the routes for the logged-out state
    const routes = [
      { name: "Login", href: "/login" },
      { name: "Homepage", href: "/" },
      { name: "Users", href: "/users" },
      { name: "Trending", href: "/hashtag" },
    ];

    routes.forEach((route) => {
      const routeLink = screen.getByText(route.name);
      expect(routeLink).toBeInTheDocument();
      expect(routeLink.closest("a")).toHaveAttribute("href", route.href);
    });
  });
});
