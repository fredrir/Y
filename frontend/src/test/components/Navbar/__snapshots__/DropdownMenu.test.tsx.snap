// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownMenu > matches the snapshot 1`] = `
<div>
  <div
    class="relative block lg:hidden"
  >
    <div
      class="cursor-pointer transition-transform duration-300"
    >
      <div
        class="transition-transform duration-300 "
      >
        <svg
          aria-hidden="true"
          class="size-8"
          data-slot="icon"
          data-testid="bars-icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
    </div>
    <div
      class="absolute right-0 top-12 z-50 mt-2 w-48 overflow-hidden rounded-md border border-gray-500 bg-gray-300 text-gray-900 shadow-lg transition-all duration-300 dark:bg-gray-800 dark:text-white pointer-events-none scale-95 opacity-0"
    >
      <a
        class="flex flex-row items-center gap-2 p-2 text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
        href="/project2/user/testuser"
      >
        <svg
          aria-hidden="true"
          class="size-5"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.75 6a3.75 3.75 0 1 1-7.5 0 3.75 3.75 0 0 1 7.5 0ZM4.501 20.118a7.5 7.5 0 0 1 14.998 0A17.933 17.933 0 0 1 12 21.75c-2.676 0-5.216-.584-7.499-1.632Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        Profile
      </a>
      <a
        class="flex flex-row items-center gap-2 p-2 text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
        href="/project2"
      >
        <svg
          aria-hidden="true"
          class="size-5"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m2.25 12 8.954-8.955c.44-.439 1.152-.439 1.591 0L21.75 12M4.5 9.75v10.125c0 .621.504 1.125 1.125 1.125H9.75v-4.875c0-.621.504-1.125 1.125-1.125h2.25c.621 0 1.125.504 1.125 1.125V21h4.125c.621 0 1.125-.504 1.125-1.125V9.75M8.25 21h8.25"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        Homepage
      </a>
      <a
        class="flex flex-row items-center gap-2 p-2 text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
        href="/project2/users"
      >
        <svg
          aria-hidden="true"
          class="size-5"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 19.128a9.38 9.38 0 0 0 2.625.372 9.337 9.337 0 0 0 4.121-.952 4.125 4.125 0 0 0-7.533-2.493M15 19.128v-.003c0-1.113-.285-2.16-.786-3.07M15 19.128v.106A12.318 12.318 0 0 1 8.624 21c-2.331 0-4.512-.645-6.374-1.766l-.001-.109a6.375 6.375 0 0 1 11.964-3.07M12 6.375a3.375 3.375 0 1 1-6.75 0 3.375 3.375 0 0 1 6.75 0Zm8.25 2.25a2.625 2.625 0 1 1-5.25 0 2.625 2.625 0 0 1 5.25 0Z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        Users
      </a>
      <a
        class="flex flex-row items-center gap-2 p-2 text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
        href="/project2/hashtag"
      >
        <svg
          aria-hidden="true"
          class="size-5"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.25 8.25h15m-16.5 7.5h15m-1.8-13.5-3.9 19.5m-2.1-19.5-3.9 19.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        Trending
      </a>
      <div
        class="hidden lg:block"
      >
        <button
          aria-label="Switch to dark theme"
          class="mx-4 h-10 w-16 rounded-full p-1 transition-colors duration-300 focus:outline-none bg-gray-300"
        >
          <div
            class="flex h-8 w-8 transform items-center justify-center rounded-full transition-transform duration-300 translate-x-0 bg-white"
          >
            <svg
              class="lucide lucide-sun h-5 w-5 text-yellow-500"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="4"
              />
              <path
                d="M12 2v2"
              />
              <path
                d="M12 20v2"
              />
              <path
                d="m4.93 4.93 1.41 1.41"
              />
              <path
                d="m17.66 17.66 1.41 1.41"
              />
              <path
                d="M2 12h2"
              />
              <path
                d="M20 12h2"
              />
              <path
                d="m6.34 17.66-1.41 1.41"
              />
              <path
                d="m19.07 4.93-1.41 1.41"
              />
            </svg>
          </div>
        </button>
      </div>
      <div
        class="lg:hidden"
      >
        <button
          class="flex w-full flex-row items-center gap-2 p-2 text-left text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
        >
          <svg
            class="lucide lucide-moon h-5 w-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
            />
          </svg>
          <p>
            Dark Mode
          </p>
        </button>
      </div>
      <button
        class="flex w-full flex-row items-center gap-2 p-2 text-left text-lg hover:bg-gray-200 dark:hover:bg-gray-900"
      >
        <svg
          class="lucide lucide-log-out size-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
          />
          <polyline
            points="16 17 21 12 16 7"
          />
          <line
            x1="21"
            x2="9"
            y1="12"
            y2="12"
          />
        </svg>
        Log out
      </button>
    </div>
  </div>
</div>
`;
